core/src/main/java/org/apache/iceberg/CachingCatalog.java

```diff
@@ -145,22 +146,26 @@ public Table loadTable(TableIdentifier ident) {
     }
 
     if (MetadataTableUtils.hasMetadataTableName(canonicalized)) {
```

Comment:
Do you suggest we call `catalog::loadTable` first? If so, we might not be able to cache the original table, since the underlying catalog can load the metadata table directly.

Reply:
What I had in mind was something different:
try {
  return tableCache.get(canonicalized, catalog::loadTable);
} catch (NoSuchTableException) {
  // Verify it's a metadata table
  // Do whatever we want in case of a metadata table, even cache the original table.
}
What do you think?

Reply:
The issue is not`NoSuchTableException` will ever be thrown on cache miss since `catalog::loadTable` can successfully load the metadata table. For example, in `BaseMetastoreCatalog`

https://github.com/apache/iceberg/blob/fa00482d3fb45a053d9569f82983295a5c6a7499/core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java#L45-L71

In this case, we don't cache the original table and change the current behavior.

Reply:
Finally I had some time to take a deeper look on this. You are right, with my above proposal we won't run the specific code for metadata tables like caching the underlying table and sharing the table operations.

I did find a way simpler version of this fix, however. I also wrote some tests in TestCachingCatalog. https://github.com/gaborkaszab/iceberg/commit/c1eca8b09e9bb8eb595fa5ce5e16e6f50636a783
What do you think? Would you mind if I opened an alternative fix PR?

Reply:
Please go ahead.

Reply:
Thanks @manuzhang !
I created this PR for the alternative fix: https://github.com/apache/iceberg/pull/11738 Added you as a co-author.
@RussellSpitzer I saw your thumbs-up on one of my comments. Would you mind taking a look at the PR?